situp-core/src/main/java/com/amazon/situp/parser/PipelineParser.java

```diff
@@ -1,32 +1,155 @@
+            final Map<String, PipelineConfiguration> pipelineConfigurationMap = OBJECT_MAPPER.readValue(
+                    new File(configurationFileLocation), new TypeReference<Map<String, PipelineConfiguration>>() {
+                    });
+            final List<String> allPipelineNames = PipelineConfigurationValidator.
+                    validateAndGetPipelineNames(pipelineConfigurationMap);
+            final Map<String, Pipeline> pipelineMap = new HashMap<>();
+            for (String pipelineName : allPipelineNames) {
+                if (!pipelineMap.containsKey(pipelineName)) {
+                    buildPipelineFromConfiguration(pipelineName, pipelineConfigurationMap, pipelineMap);
+                }
+            }
+            return pipelineMap;
         } catch (IOException e) {
             throw new ParseException(format("Failed to parse the configuration file %s", configurationFileLocation), e);
         }
     }
-}
+
+    private void buildPipelineFromConfiguration(
+            final String pipelineName,
+            final Map<String, PipelineConfiguration> pipelineConfigurationMap,
+            final Map<String, Pipeline> pipelineMap) {
+        final PipelineConfiguration pipelineConfiguration = pipelineConfigurationMap.get(pipelineName);
+
+        final PluginSetting sourceSetting = pipelineConfiguration.getSource().getPluginSettings().get(0);
+        final Optional<Source> pipelineSource = getSourceIfPipelineType(pipelineName, sourceSetting,
+                pipelineMap, pipelineConfigurationMap);
+        final Source source = pipelineSource.orElseGet(() -> SourceFactory.newSource(sourceSetting));
+
+        final Buffer buffer = getBufferOrDefault(pipelineConfiguration.getBuffer());
+
+        final Configuration processorConfiguration = pipelineConfiguration.getProcessor();
+        final List<PluginSetting> processorPluginSettings = processorConfiguration.getPluginSettings();
+        final List<Processor> processors = processorPluginSettings.stream()
+                .map(ProcessorFactory::newProcessor)
+                .collect(Collectors.toList());
+        final int processorThreads = getWorkersOrDefault(processorConfiguration);
+        final int readBatchDelay = getDelayOrDefault(processorConfiguration);
+
+        final List<PluginSetting> sinkPluginSettings = pipelineConfiguration.getSink().getPluginSettings();
+        final List<Sink> sinks = sinkPluginSettings.stream().map(this::buildSinkOrConnector).collect(Collectors.toList());
+
+        final Pipeline pipeline = new Pipeline(pipelineName, source, buffer, processors, sinks, processorThreads, readBatchDelay);
+        pipelineMap.put(pipelineName, pipeline);
+    }
+
+    private Buffer getBufferOrDefault(final Configuration bufferConfiguration) {
+        final List<PluginSetting> pluginSettings = bufferConfiguration.getPluginSettings();
+        return pluginSettings.isEmpty() ? new BlockingBuffer() : BufferFactory.newBuffer(pluginSettings.get(0));
```

Comment:
The `.get(0)` can be dangerous.

Reply:
basically we allow only one Buffer to be defined and if PluginSetting is not empty, it means it has only one entry [validation does the check before reaching this point]

Reply:
basically i mean `get(0)` is safe here

Reply:
I agree the validation to have one buffer is done before this getBufferOfDefault() but I think we should not return a list if thats so. As discussed, we should look at making this structure less dangerous.